tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,743 @@
+//    the value;
+// 2. most of types derive from llvm::User which has "operands()" method to get
+//    other llvm::Value nodes being used by the value;
+//
+// [Limitation]
+// For now the search doesn't go beyond the function boundary because the
+// reference to op name string literals and c10 op registration/invocation
+// APIs are almost always in the same function. If we create helper function
+// around c10 API (e.g. the "callOp" method defined in aten/native), we could
+// simply add them to the regular expression used to identify c10 API.
+//
+// [Example]
+// In the following example, it finds out:
+//  1) the registered function for "quantized:add" operator;
+//  2) one possible call path to at::empty() function;
+//  3) the called operator name "aten::empty":
+//
+// - quantized::add
+// - c10::detail::wrap_kernel_functor_unboxed_<at::native::(anonymous
+//   namespace)::QAdd<false>, at::Tensor (at::Tensor, at::Tensor, double,
+//   long)>::call(c10::OperatorKernel*, at::Tensor, at::Tensor, double, long)
+// - at::native::(anonymous namespace)::QAdd<false>::operator()(at::Tensor,
+//   at::Tensor, double, long)
+// - void at::native::DispatchStub<void (*)(at::Tensor&, at::Tensor const&,
+//   at::Tensor const&), at::native::qadd_stub>::operator()<at::Tensor&,
+//   at::Tensor const&, at::Tensor const&>(c10::DeviceType, at::Tensor&,
+//   at::Tensor const&, at::Tensor const&)
+// - at::native::DispatchStub<void (*)(at::Tensor&, at::Tensor const&,
+//   at::Tensor const&), at::native::qadd_stub>::choose_cpu_impl()
+// - void at::native::(anonymous namespace)::qadd_kernel<false>(at::Tensor&,
+//   at::Tensor const&, at::Tensor const&)
+// - at::TensorIterator::binary_op(at::Tensor&, at::Tensor const&, at::Tensor
+//   const&, bool)
+// - at::TensorIterator::build()
+// - at::TensorIterator::fast_set_up()
+// - at::empty(c10::ArrayRef<long>, c10::TensorOptions const&,
+//   c10::optional<c10::MemoryFormat>)
+// - aten::empty
+
+#include <deque>
+#include <iostream>
+#include <set>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "llvm/Demangle/Demangle.h"
+#include "llvm/Analysis/LazyCallGraph.h"
+#include "llvm/IR/CallSite.h"
+#include "llvm/IR/Constant.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Regex.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+
+namespace {
+
+struct RegexOpt {
+  std::shared_ptr<Regex> pattern;
+
+  void operator=(const std::string& val) {
+    if (val.empty()) {
+      return;
+    }
+    pattern = std::make_shared<Regex>(val);
+    std::string regexError;
+    if (!pattern->isValid(regexError)) {
+      report_fatal_error(
+          "Invalid regular expression param: '" + val + "' err: " + regexError,
+          false);
+    }
+  };
+};
+
+static RegexOpt FunctionSchemaPatternLoc;
```

Comment:
Nit: the c++ way would be to use an anonymous namespace instead of making everything static

Reply:
I agree with most of your style related comments here. Part of the reason I wrote this way was because I simply followed some existing examples in llvm, e.g.: https://github.com/llvm-mirror/llvm/blob/master/tools/llvm-opt-report/OptReport.cpp
And I just landed this PR before seeing your comments. I'll try to address them in subsequent PRs.